### PR TITLE
High light route

### PR DESCRIPTION
**Pull request by Nico281102**
# Description

Added a new feature to the Android navigation app. When a Point of Interest (POI) is clicked in the list, the app now highlights the road or route leading to that POI on the map. 

# How Has This Been Tested?

- [X] Manually
- [ ] Unit tests
- [ ] Integration tests

# Screenshots / Videos

Screenshot_20231222_172113

### CODE REVIEW

File: app/src/main/java/com/appocalypse/naturenav/map/MapViewModel.java [hunk 1]
```diff
@@ -1,10 +1,25 @@
...
     }
+
+    public LiveData<POI> getHighlightRouteRequest() {
+        return highlightRouteRequest;
+    }
+
+    public void requestHighlightRoute(POI poi) {
+        highlightRouteRequest.setValue(poi);
+    }
+
```
Comment (NeverMendel) [inline app/src/main/java/com/appocalypse/naturenav/map/MapViewModel.java:24]: Are those changes still necessary?

File: app/src/main/java/com/appocalypse/naturenav/map/MapFragment.java [hunk 2]
```diff
@@ -39,12 +40,16 @@ public class MapFragment extends Fragment {
 
     private FragmentMapBinding binding;
     private GeoPoint myLocation;
+    private MapView mapView;
+
+    private List<Marker> poiMarkers = new ArrayList<>();
+
+    public static String getTAG() {
+        return TAG;
+    }
```
Comment (NeverMendel) [inline app/src/main/java/com/appocalypse/naturenav/map/MapFragment.java:49]: Is this needed?

File: app/src/main/java/com/appocalypse/naturenav/map/MapFragment.java [hunk 3]
```diff
@@ -97,14 +103,26 @@ public View onCreateView(@NonNull LayoutInflater inflater,
...
+        poiInfoViewModel.getDisplayedPoiLiveData().observe(getViewLifecycleOwner(), poi -> {
+            if (poi != null) {
+                highlightRouteToPoi(poi);
+            }
+        });
+
+        bottomSheetDialogViewModel.getDisplayingListLiveData().observe(getViewLifecycleOwner(), isNotSelected -> {
+            if (isNotSelected) {
+                removeRouteHighlight();
+            }
```
Comment (NeverMendel) [inline app/src/main/java/com/appocalypse/naturenav/map/MapFragment.java:115]: else { highlight route }

File: app/src/main/java/com/appocalypse/naturenav/map/MapFragment.java [hunk 4]
```diff
@@ -137,4 +155,33 @@ public void onDestroyView() {
...
+            mapView.getOverlays().add(polyline);
+
+            // Refresh the map
+            mapView.invalidate();
+        }
+    }
+    public void removeRouteHighlight() {
+        mapView.getOverlays().removeIf(overlay -> overlay instanceof Polyline);
+        mapView.invalidate();
+    }
```
Comment (NeverMendel) [inline app/src/main/java/com/appocalypse/naturenav/map/MapFragment.java:186]: Would it look better if you temporarly removed all other POI markers when one is selected?

Then, when the POI is deselected, all POI markers would be restored and displayed again

File: app/src/main/java/com/appocalypse/naturenav/poilist/PoiListFragment.java [hunk 5]
```diff
@@ -13,22 +13,22 @@
...
     private final PoiListAdapter adapter = new PoiListAdapter();
+    private FragmentPoiListBinding binding;
 
     public View onCreateView(@NonNull LayoutInflater inflater,
                              ViewGroup container, Bundle savedInstanceState) {
         PoiListViewModel listViewModel =
                 new ViewModelProvider(requireActivity()).get(PoiListViewModel.class);
         BottomSheetDialogViewModel bottomSheetDialogViewModel = new ViewModelProvider(requireActivity()).get(BottomSheetDialogViewModel.class);
         PoiInfoViewModel poiInfoViewModel = new ViewModelProvider(requireActivity()).get(PoiInfoViewModel.class);
+        MapViewModel mapViewModel = new ViewModelProvider(requireActivity()).get(MapViewModel.class);
```
Comment (NeverMendel) [inline app/src/main/java/com/appocalypse/naturenav/poilist/PoiListFragment.java:31]: Are those changes still necessary?

File: app/src/main/java/com/appocalypse/naturenav/map/MapFragment.java (hunk 3 above)
Comment (Nico281102) [reply app/src/main/java/com/appocalypse/naturenav/map/MapFragment.java:115]: I don't have the POI object in that case, the highlitghtroute function needs the POI

File: app/src/main/java/com/appocalypse/naturenav/map/MapFragment.java (hunk 4 above)
Comment (Nico281102) [reply app/src/main/java/com/appocalypse/naturenav/map/MapFragment.java:182]: It could be better if we show the marker of the poi selected on the map?

Comment (NeverMendel) [reply app/src/main/java/com/appocalypse/naturenav/map/MapFragment.java:182]: Yep, only that one though

**Review by NeverMendel — APPROVED** [review]
Fixed a bug: POI still displaying when bottom sheet dialog is closed

LGTM now

**Pull request merged**